---
title: "Thursday.rmd"
author: "Corinne Strawser & Matt Gust"
date: "2025-05-08"
output: html_document
---

# Session Description

This session will provide an introduction to differential expression (DE) analysis and gene set enrichment analysis (GSEA) of single-cell RNA-seq data. This session will also cover experimental design considerations.

# Session Learning Objectives

After completion, you should be able to perform DE and GSEA analysis using scRNA-seq and pseudobulk methods. You should also be able to visualize the output of your DE and GSEA analyses.

# Section 1: Prepare for Differential Expression Testing

* Load in integrated aging murine brain Seurat object from Wednesday
* Visualize data as a UMAP grouped by cell type and split by condition
* Tally the number of cells in each cell type per condition
* Tally the number of cells in each cell type per condition and sample

------------------------------------------------------------------------------------

# Section 2: Introduction to Differential Expression (DE) Analysis

* Differential gene expression between clusters
  * FindAllMarkers
    * Visualize markers on a heatmap
    * Visualize markers on a FeaturePlot
  * FindConservedMarkers
* **Activity 1**
  * Change identity to cell type
  * Visualize cell types as a UMAP split by condition
  * Run FindAllMarkers to identify genes defining each cell type
  * Find cell type markers and visualize as dotplot
* Differential gene expression within cell types between conditions
  * Find genes differentially expressed in a cluster between conditions
  * Explore the results
  * Save DEGs as a .csv file
  * Visualize data as a volcano plot
  * Visualize select genes as a FeaturePlot
  * Visualize select genes as a VlnPlot
  * Finding DEGs in a for-loop
* **Activity 2**
  * Find genes that are differentially expressed between oligodendrocytes between old and young mice
  * Explore results and save as a .csv file
  * Visualize top 3 most significantly different genes using VlnPlot
* Pseudobulked Differential Gene Expression
  * Why pseudobulk?
  * Prepare data for pseudobulk analysis
  * Perform DESeq2 on pseudobulk data
    * Save DEGs as a .csv file
    * Visualize data as a volcano plot
* **Activity 3**
  * Use the same cell type as in Activity 2
  * Find genes that are differentially expressed between two conditions using the pseudobulk method
  * Explore results and save as a .csv file
  * Visualize top 3 most significantly different genes using VlnPlot
  * How do the results from the two DEG analyses compare?

------------------------------------------------------------------------------------

# Section 3: Introduction to Gene Set Enrichment Analysis (GSEA)

* Perform GSEA analysis on a list of DEGs with fgsea
  * Save GSEA results as a .csv file
  * Visualize data as a bubble plot
* **Activity 4**
  * Perform GSEA analysis on the astrocytes old vs. young DEGs using the Gene Ontology Biological Processes geneset
  * Visualize GSEA results as a bubble plot

------------------------------------------------------------------------------------

# Section 4: Experimental Design Considerations


# Set up

```{r setup, include=FALSE}
library(Seurat)
library(tidyverse)
library(msigdbr)
library(fgsea)
library(data.table)
library(RColorBrewer)
library(ggrepel)
library(DESeq2)
library(qs)

# set seed for reproducibility
set.seed(888)

# set directories
session_path <- "/igm/home/cnh008/projects/scrgot/scrgot_2025/NCH_Coder_Upgrade/Sessions/beginner_thursday/"
data_path <- "/igm/home/cnh008/projects/scrgot/scrgot_2025/workshop_data"
setwd(session_path)
```

# Section 1: Prepare for Differential Expression Testing

Load in integrated aging murine brain Seurat object from Wednesday

```{r load_data}
data <- 
  qread(file = paste0(data_path, "/rdata_premade/brain_scRNA_combined.qs"))

# First, make sure that our data layers are joined
data@assays$RNA@layers %>% 
  names()
# if it's not joined you'll have several counts and data layers

# join layers after integration if not 
# data[["RNA"]] <- 
#   JoinLayers(data[["RNA"]])

# check / set default assay
DefaultAssay(data) <- "RNA"

# Second, check to make sure relevant meta data is in object
data@meta.data %>%
  head()

# if not, add it in
## add in an age variable
data$age <-
  ifelse(test = str_detect(string = data$orig.ident, pattern = "OX"),
         yes = "Old",
         no = "Young")
## add in a mouse variable
data$mouse <-
  str_split_fixed(string = data$orig.ident,
                  pattern = "_",
                  n = 2)[,2]
data@meta.data %>%
  head()

# set identity for the dataset
Idents(data) <- 
  "cell_type"
```

Look at the distribution of cells in each cell type when grouped by age or mouse

```{r visualize_data}
# plot UMAP separated by cell type, age, and mouse
DimPlot(object = data,
        group.by = c("cell_type", "age", "mouse"))
```

Tally the number of cells in each cell type per age group

```{r tally_cells}
# how many cells do we have per age group?
table(data$cell_type, data$age)
```

Tally the number of cells in each cell type per condition and sample

```{r tally_cells_2}
# how many cells do we have per mouse?
table(data$cell_type, data$age, data$mouse)
```

--------------------------------------------------------------------------------

# Section 2: Introduction to Differential Expression (DE) Analysis

In this section we will learn how to perform differential expression analysis using Seurat built-in functions. We will use the RNA assay.
```{r}
# check to see if RNA data has been normalized yet
GetAssayData(object = data,
             assay = "RNA",
             layer = "data")[1:10,1:10]
# integers typically indicate data has not been normalized
# floats typically indicate data has been normalized
```

## Between clusters

We can use any slot in the metadata for performing DE testing. We will start with clusters.

### Find markers that define clusters with FindAllMarkers

First, we will confirm that our identity is set to our rpca_cluster metadata slot. We will then perform DE testing for all clusters in our data using the FindAllMarkers function which will identify genes that are differentially expressed in each identity group by comparing it to all of the others. Note that this method does not identify markers that are specific to a single group, and thus, closely-related groups may have overlapping markers.

```{r FindAllMarkers}
# make sure identity is set to seurat clusters
Idents(data) <- "seurat_clusters"
levels(data) %>%
  unique()

# look at umap
DimPlot(object = data,
        label = T,
        label.box = T,
        repel = T)

# find markers for each cluster
markers <- 
  FindAllMarkers(object = data,
                 min.pct = 0.25,
                 logfc.threshold = 0.25,
                 only.pos = T)
# this finds genes that differentially expressed for each cluster selecting
# for genes that are expressed in >25% of cells and only considering genes
# that have a positive log2FC > 0.25 

# reduce data
top_markers <- 
  markers %>%
  as.data.frame() %>%
  group_by(cluster) %>% # group by cluster
  slice_min(order_by = p_val_adj, 
            n = 3,
            with_ties = F) # sort by adjusted p-value and get bottom 3

# explore top markers
head(top_markers)

# save markers as a dataframe
write.csv(x = top_markers, 
          file = "FindAllMarkers_top3_cluster_markers.csv")
```

### Visualize markers as a heatmap

Sometimes it's nice to see globally which markers are defining specific groups and a heatmap is a great way to do that. We will want to use normalized counts for this so that each group has comparable values. This will show the normalized expression for each gene for each cell in each identity.

```{r}
# get list of top marker genes; including each gene only once
genes <- 
  top_markers$gene %>%
  unique()

# change to RNA assay
DefaultAssay(data) <- "RNA"

# scale data for visualization
data <-
  ScaleData(object = data,
            assay = "RNA")

# make heatmap
small <-
  subset(x = data,
         downsample = 50)
DoHeatmap(object = small,
          features = genes,
          group.by = "seurat_clusters")
ggsave(filename = "top3_cluster_markers.jpeg",
       width = 8,
       height = 8)
```

### Visualize markers as a FeaturePlot

```{r}
# plot some top gene per cluster
top_markers <- 
  markers %>%
  as.data.frame() %>%
  group_by(cluster) %>% # group by cluster
  slice_min(order_by = p_val_adj, 
            n = 1,
            with_ties = F) # sort by adjusted p-value and get bottom 1
genes <- 
  top_markers$gene

# make FeaturePlot
FeaturePlot(object = data,
            features = genes,
            ncol = 5,
            label = T)
```

### Find markers that define clusters with FindConservedMarkers

We can use FindConservedMarkers to identify genes that are conserved between the conditions for each identity class and are differentially expressed compared to other identity classes. This is helpful to identify true marker genes for each cluster. Often the results of FindConservedMarkers will be similar to FindAllMarkers. We won't run it here but it's something to be aware of.

```{r FindConservedMarkers, eval = FALSE}
conserved <- 
  FindConservedMarkers(data,
                       ident.1 = "0",
                       grouping.var = "age")
```

--------------------------------------------------------------------------------

## Activity 1

### Change the identity class to *cell type* and make a UMAP colored by cell type and split by age
```{r}
# make sure identity is set to cell types
Idents(data) <- "cell_type"
levels(data) %>%
  unique()

# make UMAP
DimPlot(data,
        split.by = "age")
```

### Use FindAllMarkers to identify genes defining each cell type
```{r}
# find markers for each cell type
markers <- 
  FindAllMarkers(object = data,
                 min.pct = 0.25,
                 logfc.threshold = 0.25,
                 only.pos = T)
# this finds genes that differentially expressed for each cell type selecting
# for genes that are expressed in >25% of cells and only considering genes
# that have a positive log2FC > 0.25

head(markers)
```

### Make a DotPlot for the top 2 markers per cell type (hint: how do you get information for how to run a function?)

Why might a dot plot be useful for showing scRNA-seq data?
```{r}
# reduce data
top_markers <- 
  markers %>%
  as.data.frame() %>%
  group_by(cluster) %>% # group by cluster (cell type)
  slice_max(order_by = avg_log2FC, 
            n = 2,
            with_ties = F) # sort by log2FC and get top 2

# get genes
genes <-
  top_markers$gene %>%
  unique()

# make dot plot
DotPlot(object = data,
        features = genes) +
  RotatedAxis() +
  scale_color_continuous(type = "viridis")

# A dot plot is a nice visualization for scRNA-seq data because it clearly displays the percent of cells that express the gene and the level of expression
```

--------------------------------------------------------------------------------

## Within clusters between conditions

If we have two conditions, we likely want to compare them! To do that, we can use the FindMarkers function.

### Find genes differentially expressed in one cluster between conditions

Here we will find the genes that are differentially expressed between a cluster shared between young and old mouse brains. We want genes that are expressed in at least 10% of the cells in each identity and have a log2FC cut-off as 0.25 (1.2X).
```{r}
# how many cells do we have per condition?
table(data$seurat_clusters, data$age)

# change identity class
Idents(data) <- "seurat_clusters"

# set assay
DefaultAssay(data) <- "RNA"

# Find DEGs for cluster 0
degs <- 
  FindMarkers(object = data,
              ident.1 = "Old",
              ident.2 = "Young",
              group.by = "age",
              subset.ident = "0",
              min.pct = 0.1,
              logfc.threshold = 0.25)
```

### Explore the results
```{r}
# look at first 20 differentially expressed genes
degs %>%
  head(n = 20)

# how many significantly different genes are there?
degs[degs$p_val_adj <= 0.05,] %>%
  nrow()
```

The results data frame has the following columns :

* p_val : p_val (unadjusted)
* avg_log2FC : log fold-change of the average expression between the two groups. 
** Note : positive values indicate that the feature is more highly expressed in the first group.
* pct.1 : percentage of cells where the feature is detected in the first group
* pct.2 : percentage of cells where the feature is detected in the second group
* p_val_adj : adjusted p-value, based on Bonferroni correction using all features in the dataset.

### Save DEGs as a .csv file

We usually want to export the files to share with others and saving them as 
a .csv file is a great way to do that.

```{r}
write.csv(x = degs, 
          file = "cluster_0_old_vs_young_degs.csv")
```

### Visualize data as a volcano plot

A volcano plot allows us to see gene expression changes and their significance 
for a lot of genes at once.

```{r}
# log transform the padj values for visualization
degs$log.padj <- 
  -log10(degs$p_val_adj)

# label whether the genes were significant
degs$significant <- 
  ifelse(degs$p_val_adj <= 0.05,
         "p < 0.05",
         "p > 0.05")

# pull out labels for the top 20 most differentially expressed genes
labels <- 
  degs %>%
  arrange(desc(abs(avg_log2FC))) %>%
  slice_head(n = 20)

# make volcano plot
volcano <- 
  ggplot(data = degs, 
         aes(x = avg_log2FC, 
             y = log.padj,
             color = significant)) +
  geom_point() +
  scale_color_manual(values = c("darkmagenta", "grey")) +
  geom_label_repel(data = labels,
                   aes(label = rownames(labels)),
                   max.overlaps = 20,
                   show.legend = FALSE) +
  ggtitle("Cluster 0 Differentially Expressed Genes\nOld vs. Young Mouse") +
  ylab("-log(padj)") +
  xlab("log2FC") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        text = element_text(size = 12))
volcano
```

### Visualize select genes as a FeaturePlot

A feature plot can allow us to map gene expression onto our UMAP. We can also 
split it based on a slot in the metadata, such as treatment.

```{r}
# make feature plot of one down-regulated and two up-regulated genes from volcano plot
FeaturePlot(object = data,
            features = c("Malat1", "Ogt"),
            split.by = "age",
            label = T,
            order = T) & # pulls positive cells to the front
  scale_color_continuous(type = "viridis")
```

### Visualize select genes as a VlnPlot

A violin plot allows us to see gene expression levels within each identity.

```{r}
# make violin plot
VlnPlot(object = data,
        features = c("Malat1"),
        split.by = "age",
        group.by = "seurat_clusters",
        cols = c("#0072B2","#CC79A7"),
        ncol = 1,
        pt.size = 0) # removes the black dots for each cell
```

### Intro to for-loops in R

In your dataset you'll likely have a few if not a dozen or more clusters and/or cell types. Thus, instead of writing out code to do the same thing over and over (like find DEGs) it is useful and more reproducible to iterate over elements in a list, dataframe, vector, matrix, etc. The for loop can be used to execute a group of statements or functions repeatedly depending on the number of elements in your object. 

```{r}
# Example 1
fruits <- c("apple", "banana", "cherry")

for (x in fruits) {
  print(x)
}

# Example 2
my_fav_nums <- c(4,5,6)

for (x in my_fav_nums){
  y <- x + 1
  print(y)
}
```

We could also find DEGs for several clusters at a time using a for-loop

```{r}
# select only clusters that have >3 cells in both groups
cells_per_group <- 
  table(data$seurat_clusters, 
        data$age) %>%
  as.data.frame() %>%
  pivot_wider(names_from = Var2,
              values_from = Freq) %>%
  as.data.frame()
head(cells_per_group)
clusters <-
  cells_per_group %>%
  filter(Old >= 3 & Young >= 3) %>%
  pull(Var1) %>%
  as.character()

# Find DEGs for each cluster
for (cell in clusters){
    # Find the markers
    markers <- 
      FindMarkers(object = data,
                  ident.1 = "Old",
                  ident.2 = "Young",
                  group.by = "age",
                  subset.ident = cell,
                  min.pct = 0.1,
                  logfc.threshold = 0.25)
  
    # make a variable for a filename
    filename <- 
      paste0("cluster_",cell, "_old_vs_young_degs.csv")
    
    # save our DEGs
    write.csv(markers, 
              file=filename)
}
```

--------------------------------------------------------------------------------

## Activity 2

### Find genes that are differentially expressed between oligodendrocytes between old and young mice

```{r}
# how many cells do we have per condition?
table(data$cell_type, data$age)

# change identity class
Idents(data) <- "cell_type"

# set assay
DefaultAssay(data) <- "RNA"

# Find DEGs for cluster oligodendrocytes
degs <- 
  FindMarkers(object = data,
              ident.1 = "Old",
              ident.2 = "Young",
              group.by = "age",
              subset.ident = "Oligo",
              min.pct = 0.1,
              logfc.threshold = 0.25)
```

### Explore results and save as a .csv file. How many significant degs are there?

```{r}
head(degs,
     n = 20)

write.csv(x = degs, 
          file = "oligos_old_vs_young_degs.csv")

degs[degs$p_val_adj <= 0.05,] %>%
  nrow()
# 494
```

### Visualize top 3 most significantly different genes in oligodendrocytes by p_val_adj using VlnPlot (hint: how did we subset to get a small dataset? instead of downsample, what could we use?)

```{r}
genes <-
  degs %>%
  slice_min(order_by = p_val_adj,
            n = 3,
            with_ties = F) %>%
  rownames()

VlnPlot(subset(data, cell_type == "Oligo"),
        features = genes,
        split.by = "age",
        group.by = "cell_type",
        cols = c("#0072B2","#CC79A7"),
        ncol = 3)
```

--------------------------------------------------------------------------------
  
# Pseudobulked Differential Gene Expression

https://satijalab.org/seurat/articles/de_vignette

## Why pseudobulk?
The DE analysis in Seurat treats each cell as an independent sample often resulting in inflated p-values as single cells isolated from the same biological sample are not *truly* independent of one another. Thus, the methods used to perform DE analysis in Seurat identify genes that vary across individual cells not necessarily across a population. To infer DE genes across a population between conditions we need to pseudobulk our cells.

## Important note:
In order to perform pseudobulked DEG analysis you should have at least 3 biological replicates.

## Prepare data for pseudobulk analysis
First, generate DEG tables from pseudobulk clusters, treating all cells from each ident from each sample as independent groups of cells/data. Then, graph each of those DEG datasets using a volcano plot.

```{r}
# what do our cells look like before pseudobulking?
Cells(data) %>%
  head()

# create pseudobulked data for clusters across condition
pb_data <- 
  AggregateExpression(object = data, 
                      assays = "RNA", 
                      return.seurat = T, 
                      group.by = c("age", 
                                   "mouse", 
                                   "seurat_clusters"))

# what do our cells look like after pseudobulking?
Cells(pb_data) %>%
  head()
# cage_mouse_cluster

# add a column to the metadata that is cluster_age
pb_data$cluster_age <- 
  paste(pb_data$seurat_clusters,
        pb_data$age,
        sep = "_")
pb_data$cluster_age %>%
  head()

# how many samples do we have at each level?
pb_data$cluster_age %>%
  table()

# set default assay to RNA
DefaultAssay(pb_data) <- "RNA"

# perform DE testing on pseudbulked data for same cluster as above
Idents(pb_data) <- "cluster_age"
pseudobulked_degs <- 
  FindMarkers(object = pb_data, 
              ident.1 = "0_Old",
              ident.2 = "0_Young",
              test.use = "DESeq2")
pseudobulked_degs %>%
  head(n = 10)

# how many significant DEGs do we have?
pseudobulked_degs[pseudobulked_degs$p_val_adj <= 0.05,] %>%
  nrow()
```

### Visualize data as volcano plots
```{r}
# log transform the padj values for visualization
pseudobulked_degs$log.padj <- 
  -log10(pseudobulked_degs$p_val_adj)

# label whether the genes were significant
pseudobulked_degs$significant <- 
  ifelse(pseudobulked_degs$p_val_adj <= 0.05,
         "p < 0.05",
         "p > 0.05")

# pull out labels for the top 20 most differentially expressed genes
labels <- 
  pseudobulked_degs %>%
  arrange(desc(abs(avg_log2FC))) %>%
  slice_head(n = 20)

# make volcano plot
volcano <- 
  ggplot(data = pseudobulked_degs, 
         aes(x = avg_log2FC, 
             y = log.padj,
             color = significant)) +
  geom_point() +
  scale_color_manual(values = c("darkmagenta", "grey")) +
  geom_label_repel(data = labels,
                   aes(label = rownames(labels)),
                   max.overlaps = 20,
                   show.legend = FALSE) +
  ggtitle("Cluster 0 Differentially Expressed Genes\nOld vs. Young Mice") +
  ylab("-log(padj)") +
  xlab("log2FC") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5),
        text = element_text(size = 12))
volcano
```

## Activity 3:

### Create a pseudobulk dataset for *cell types*
```{r}
# create pseudobulked data for clusters across condition
pb_data <- 
  AggregateExpression(object = data, 
                      assays = "RNA", 
                      return.seurat = T, 
                      group.by = c("age", 
                                   "mouse", 
                                   "cell_type"))

# add a column to the metadata that is cell_type_age
pb_data$cell_type_age <- paste(pb_data$cell_type,
                                   pb_data$age,
                                   sep = "_")
pb_data$cell_type_age %>%
  head()
```

### Find differentially expressed genes for oligodendrocytes using the pseudobulk method
```{r}
# set default assay to RNA
DefaultAssay(pb_data) <- "RNA"

# perform DE testing on pseudbulked data for same cell type as above
Idents(pb_data) <- "cell_type_age"
pseudobulked_degs <- 
  FindMarkers(object = pb_data, 
              ident.1 = "Oligo_Old",
              ident.2 = "Oligo_Young",
              test.use = "DESeq2")
pseudobulked_degs %>%
  head(n = 10)

# how many significant DEGs do we have?
pseudobulked_degs[pseudobulked_degs$p_val_adj <= 0.05,] %>%
  nrow()
# 252
```

### How many FEWER DEGs do we have with the pseudobulk method? Compare a gene that is changed in the same direction in both and a gene that is only in the pseudobulk method visually. How do they compare?
```{r}
# How many fewer?
nrow(degs[degs$p_val_adj <= 0.05,]) - nrow(pseudobulked_degs[pseudobulked_degs$p_val_adj <= 0.05,]) 
# 242

intersect(rownames(degs), 
          rownames(pseudobulked_degs)) %>%
  head()

setdiff(rownames(pseudobulked_degs),
        rownames(degs)) %>%
  head()

genes <- c("C4b", "Cdkn2c")
Idents(data) <- "cell_type"
oligo <- subset(data, idents = "Oligo")
VlnPlot(object = oligo,
        features = genes,
        group.by = "age")
```

------------------------------------------------------------------------------------

# Section 3: Introduction to Gene Set Enrichment Analysis (GSEA)

## Perform GSEA analysis on the cluster 0 differentially expressed genes

Gene set enrichment analysis (GSEA) helps us to see if several genes in a particular process or pathway are changing together. We will use the Gene Ontology Molecular Functions gene set from MSIGDB and the fgsea package. There are several GSEA packages and gene sets you can use. You can even create your own gene sets!

```{r}
Idents(data) <- "seurat_clusters"
degs <- 
  FindMarkers(object = data,
              ident.1 = "Old",
              ident.2 = "Young",
              group.by = "age",
              subset.ident = "0",
              min.pct = 0.1,
              logfc.threshold = 0.25)

# look at genesets available
View(msigdbr_collections())

# select dataset and then aggregate the genes into a list where each element of the list will be a named character vector. the name of the vector is the gene set name ("gs_name"). the elements of the vector are genes in the gene set ("gene_symbol")

go_mf <- 
  msigdbr(species = "Mus musculus",
          category = "C5",
          subcategory = "GO:MF") %>% 
  split(x = .$gene_symbol,
        f = .$gs_name)
View(go_mf)

# arrange DEGs for GSEA 
degs$gene <- 
  row.names(degs)
deg <- 
  degs %>%
  arrange(desc(avg_log2FC)) %>% 
  select(gene, avg_log2FC)

# convert to named list 
ranks <- 
  deframe(deg) 
head(ranks)

# perform GSEA
fgsea_res <- 
  fgseaMultilevel(go_mf,
                  ranks,
                  minSize = 15, # min size of gene set to test
                  maxSize = 500, # max size of a gene set to test
                  nPerm = 1000) %>% # number of permutations to test
  arrange(padj)

# take a look at results
head(fgsea_res)

# save results as a .csv file (use fwrite to handle leadingEdge list)
fwrite(fgsea_res, "cluster0_old_vs_young_GOMF_GSEA.csv")
```

The GSEA results dataframe has the following columns:

* pathway -- name of the pathway
* pval -- an enrichment p-value
* padj -- a BH-adjusted p-value
* log2err -- the expected error for the standard deviation of the P-value
* ES -- enrichment score, same as in Broad GSEA implementation
* NES -- ES normalized to mean enrichment of random samples of the same size
* size -- size of the pathway after removing genes not present
* leadingEdge -- vector with indexes of leading edge genes that drive the 
                    enrichment

### Visualize data as a bubble plot

It's always nice to display our data as a graph! We will incorporate information about the pathways enriched, the adjusted p-value, the size of the pathway, and the normalized enrichment score into a bubble plot.

```{r}
# get only the top 10 pathways
fgsea_plot <- 
  fgsea_res[1:10,]
# make bubble plot
ggplot(fgsea_plot,
       aes(x = NES,
           y = pathway,
           size = size,
           color = padj)) +
  geom_point() +
  scale_color_gradient2(high = "#f517d3",
                        mid = "#ffffff",
                        low = "#0513d8",
                        midpoint = 0.05) +
  theme_bw() +
  ggtitle("Cluster 0 Old vs. Young\nGene Ontology Molecular Function GSEA")+
  theme(plot.title = element_text(hjust = 0.5))
```

## Activity 4

Perform GSEA on the genes that are differentially expressed in the astrocytes between young and old mice

### Find genes differentially expressed in astrocytes between old and young mice
```{r}
Idents(data) <- "cell_type"
degs <- 
  FindMarkers(object = data,
              ident.1 = "Old",
              ident.2 = "Young",
              group.by = "age",
              subset.ident = "Astro",
              min.pct = 0.1,
              logfc.threshold = 0.25)
```

### Explore the results and save as a .csv file
```{r}
write.csv(degs,
          "astrocytes_old_vs_young_degs.csv")
```

### Visualize top 3 most significantly different genes using a VlnPlot
```{r}
top_markers <- 
  degs %>%
  as.data.frame() %>%
  slice_min(p_val_adj, 
            n = 3,
            with_ties = F) # sort by adjusted p-value and get bottom 3

genes <- 
  rownames(top_markers)

# make FeaturePlot
FeaturePlot(object = data,
            features = genes,
            split.by = "age",
            order = T)
```

### Visualize top 3 most significantly different genes using VlnPlot
```{r}
# make VlnPlot
VlnPlot(object = subset(data, cell_type == "Astro"),
        features = genes,
        split.by = "age")
```

## Perform GSEA analysis on the astrocyte young vs. old DEGs using the GO:BP geneset
```{r}
# this gets your geneset
go_bp <- 
  msigdbr(species = "Mus musculus",
          category = "C5",
          subcategory = "GO:BP") %>% 
  split(x = .$gene_symbol,
        f = .$gs_name)

# arrange DEGs for GSEA 
degs$gene <- 
  row.names(degs)
deg <- 
  degs %>%
  arrange(desc(avg_log2FC)) %>% 
  select(gene, avg_log2FC)

# convert to named list 
ranks <- 
  deframe(deg) 
head(ranks)

# perform GSEA
fgsea_res <- 
  fgseaMultilevel(go_bp,
                  ranks,
                  minSize = 15, # min size of gene set to test
                  maxSize = 500, # max size of a gene set to test
                  nPerm = 1000) %>% # number of permutations to test
  arrange(padj)

# take a look at results
head(fgsea_res)

# save results as a .csv file (use fwrite to handle leadingEdge list)
fwrite(fgsea_res, "astrocytes_old_vs_young_GSEA.csv")
```

### Visualize GSEA results as a bubble plot
```{r}
# get only the top 10 pathways
fgsea_plot <- 
  fgsea_res[1:10,]
# make bubble plot
ggplot(fgsea_plot,
       aes(x = NES,
           y = pathway,
           size = size,
           color = padj)) +
  geom_point() +
  scale_color_gradient2(high = "#f517d3",
                        mid = "#ffffff",
                        low = "#0513d8",
                        midpoint = 0.05) +
  theme_bw() +
  ggtitle("Astrocytes Old vs. Young\nGene Ontology Biological Processes GSEA")+
  theme(plot.title = element_text(hjust = 0.5))
```

# Section 4: Experimental Design Considerations

1. Do I really need to do scRNA-seq?
  a. single cell type of interest - probably not
  b. landscape of cell types in a sample - probably
2. What type of protocol do I need to use? Considerations:
  a. species of interest (reference species vs. unusual organism)
  b. transcript capture (poly-A 3' GEX, poly-A 5' GEX, probe-based)
  c. number of samples (multiplex vs. single plex)
  d. frequency of cell types of interest (FACS sorting, plate-based)
  e. frequency of transcripts of interest (droplet-based, plate-based)
  f. ability to identify cell types of interest (protein + GEX assays)
3. How much money do I have?
  a. capital investment (10X Chromium vs. PIP-seq vs. plate-based)
  b. number of samples
  c. sequencing requirements
  
https://satijalab.org/howmanycells/






