---
title: "Human Osteosarcoma Primary - Create datasets"
subtitle: "Cell type annotation reference assembly"
output: html_document
---

This script compiles several single cell RNA-seq data sets that are published and
combines them into a single dataset for downstream use in creating cell type references
for automated annotation. The primary output file from this script (comb.qs) is used
in several other scripts as the input to perform clustering and annotation.

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  dev = "png",
  fig.width = 6,
  fig.height = 6,
  fig.asp = NULL
)

library(rrrSingleCellUtils)
library(Seurat)
library(ggplot2)
library(ggrepel)
library(tidyverse)
library(stringr)
library(harmony)
library(cowplot)

set.seed(888)

sysfonts::font_add("Arial Narrow", "ARIALN.TTF", bold = "ARIALNB.TTF")
showtext::showtext_auto()

theme_set(ggpubr::theme_pubr())

plot_cols <- c("#D43F3AFF", "#EEA236FF", "#357EBDFF", "#5CB85CFF", "#B8B8B8FF",
  "#9632B8FF",  "#46B8DAFF", "#90302DFF", "#A66D04FF", "#2D577FFF", "#3E7E3EFF",
  "#7D7D7DFF", "#6D1D87FF", "#097F9AFF", "#FF6E6AFF", "#FFBB70FF", "#68A4E3FF",
  "#79D379FF", "#CDCDCDFF", "#BF6BE2FF", "#69D1F3FF")

rDimPlot <- function(object, title = "Unlabeled Plot",
  label = T, pt.size = 1, ...) {
  if(length(levels(Idents(object))) < 22) {  
    p <- Seurat::DimPlot(
      object = object, 
      label = label, 
      pt.size = pt.size,
      cols = alpha(plot_cols, 0.6),
      ...) +
      ggtitle(title) +
      theme(legend.position = "none") +
      coord_fixed()
    print(p)
    return(p)
  } else {
    print("Too many identity classes to use this function. Requires <22.")
  }
}

rFeaturePlot <- function(object, features, title, pt.size = 1, order = T, ...) {
  p <- Seurat::FeaturePlot(
    object = object,
    features = features,
    pt.size = pt.size,
    order = order,
    cols = (c("lightgoldenrod", "darkred")),
    ...) +
    ggtitle(title) +
    coord_fixed()
  print(p)
  return(p)
}
```

# Load published data from GSE152048

If not already done, download the data sets from the GEO repository, extract
matrices, import into a Seurat object, perform initial QC and append relevant 
clinical data for each sample, as reported in the manuscript.

```{r}
# Make a list of sample names
geo1 <- tibble(
  s = c("BC5", "BC6", "BC10", "BC11", "BC16",
    "BC17", "BC20", "BC21", "BC22"),
  qc = c(18000, 25000, 25000, 30000, 70000,
    40000, 70000, 50000, 50000),
  path = c("Conventional", "Conventional", "Conventional",
    "Conventional", "Conventional", "Chondroblastic",
    "Chondroblastic", "Intraosseous", "Chondroblastic"),
  type = c("Primary", "Primary", "Lung Met", "Primary", "Primary",
          "Lung Met", "Primary", "Primary", "Primary")
)

# Check if primary tumor directory exists and create if it doesn't
if(!dir.exists("HuOsteo/PrimaryTumor")) {
  tar_dir <- "HuOsteo/PrimaryTumor"
  dir.create(tar_dir)
}

# Download, file, and extract the files from GEO
if(!dir.exists("HuOsteo/PrimaryTumor/GSE152048")) {
  tar_dir <- "HuOsteo/PrimaryTumor/GSE152048"
  dir.create(tar_dir)
  geo_pre <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE152nnn/GSE152048/suppl/GSE152048_"
  for(s in geo1$s){
    gse_path <- str_c(geo_pre, s, ".matrix.tar.gz")
    tar_file <- str_c(tar_dir, "/", s, ".tar.gz ")
    download.file(gse_path, destfile = tar_file, method = "auto")
    untar(tar_file, exdir = tar_dir)
    file.remove(tar_file)
  }
}

# Create a vector of normalized Seurat objects for all GSE152048 samples
raw <- c()
p <- list()
for(s in geo1$s) {
  x <- tenx_load_qc(str_c("HuOsteo/PrimaryTumor/GSE152048/",
                          s, "/"))
  x <- subset(x, subset = nCount_RNA < geo1$qc[geo1$s == s] & percent.mt <13)
  x$src <- s
  x$type <- geo1$type[geo1$s == s]
  x$path <- geo1$path[geo1$s == s]
  x$gse <- "GSE152048"
  raw[[s]] <- x %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA(verbose = F) %>%
    FindNeighbors(dims = 1:20) %>%
    FindClusters(resolution = 0.3) %>%
    RunUMAP(dims = 1:20)
  p[[s]] <- rDimPlot(raw[[s]], str_c(s, " basic clustering"))
}
rm (x)
plot_grid(plotlist = p, ncol = 3)
```

# Load published data from GSE162454

Transition to data from this second manuscript and process similar to
what was done above to add these samples to the reference construct.

```{r}
# Make a list of sample names
geo2 <- tibble(
  s = c("OS_1", "OS_2", "OS_3", "OS_4", "OS_5", "OS_6"),
  qc = c(50000, 45000, 23000, 50000, 50000, 45000),
  pre = c("GSM4952363_", "GSM4952364_", "GSM4952365_", "GSM5155198_",
    "GSM5155199_", "GSM5155200_")
)

# If the raw data from GEO hasn't been downloaded, go get it
if(!dir.exists("HuOsteo/PrimaryTumor/GSE162454")) {
  tar_dir <- "HuOsteo/PrimaryTumor/GSE162454"
  dir.create(tar_dir)
  geo_pre <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE162nnn/GSE162454/suppl/GSE162454_RAW.tar"
  tar_file <- str_c(tar_dir, "/", "GSE162454.tar.gz ")
  options(timeout = 300)
  download.file(geo_pre, destfile = tar_file, method = "auto")
  untar(tar_file, exdir = tar_dir)
  file.remove(tar_file)
  for(s in geo2$s) {
    bfile <- str_c(geo2$pre[geo2$s == s], s, "_barcodes.tsv.gz")
    ffile <- str_c(geo2$pre[geo2$s == s], s, "_features.tsv.gz")
    mfile <- str_c(geo2$pre[geo2$s == s], s, "_matrix.mtx.gz")
    samp_dir <- str_c(tar_dir, "/", s)
    dir.create(samp_dir)
    file.rename(str_c(tar_dir, "/", bfile),
      str_c(samp_dir, "/", "barcodes.tsv.gz"))
    file.rename(str_c(tar_dir, "/", ffile),
      str_c(samp_dir, "/", "features.tsv.gz"))
    file.rename(str_c(tar_dir, "/", mfile),
      str_c(samp_dir, "/", "matrix.mtx.gz"))
  }
}

# Create a vector that contains normalized Seurat objects for all GSE162454
#   samples
p <- list()
for(s in geo2$s) {
  x <- tenx_load_qc(str_c("HuOsteo/PrimaryTumor/GSE162454/",
                          s, "/"))
  x <- subset(x, subset = nCount_RNA < geo2$qc[geo2$s == s] & percent.mt <18)
  x$src <- s
  x$type <- "Primary"
  x$path <- "Conventional"
  x$gse <- "GSE162454"
  raw[[s]] <- x %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA(verbose = F) %>%
    FindNeighbors(dims = 1:20) %>%
    FindClusters(resolution = 0.3) %>%
    RunUMAP(dims = 1:20)
  p[[s]] <- rDimPlot(raw[[s]], str_c(s, " basic clustering"))
}
rm(x)
plot_grid(plotlist = p, ncol = 3)

```

# Subsample and merge Seurat objects

Subsample each patient sample so that we retain a representative number of cells from
each. Merge the samples into a single Seurat object and perform a batch correction/
integration procedure using harmony. 

```{r}
# Isolate samples obtained from primary tumors
prim <- c()
raw <- raw[!names(raw) %in% c("BC10", "BC17")]

# Downsample to retain equal numbers of cells across samples (equal representation)
for(s in names(raw)) {
  prim[[s]] <- subset(raw[[s]], cells = sample(Cells(raw[[s]]), 1000))
}

# Combine the downsampled datasets into a single seurat object and re-normalize
comb <- merge(prim[[1]], y = prim[2:length(prim)],
  add.cell.ids = names(raw),
  project = "PrimaryReference"
)

comb <- comb %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(verbose = F)

# Run batch-correction/integration using harmony
comb <- RunHarmony(comb, group.by.vars = "src")
comb <- comb %>%
  RunUMAP(reduction = "harmony", dims = 1:30) %>%
  FindNeighbors(reduction = "harmony", dims = 1:30) %>%
  FindClusters()

plot <- list()
opt <- tribble(
  ~res, ~score, ~clusters
)

for(i in 1:9) {
  opt[i, "res"] <- 0.1 * i
  comb <- FindClusters(comb, resolution = opt$res[i])
  plot[[i]] <- rDimPlot(comb, paste("Resolution", opt$res[i]))
  opt[i, "score"] <- silhouette_mean(silhouette_seurat(comb))
  opt[i, "clusters"] <- length(levels(comb$seurat_clusters))
}

plot_grid(plotlist = plot, ncol = 3)
tumor <- FindClusters(comb,
 # resolution = opt$res[opt$score == max(opt$score)])
# On manual review, I am enforcing a stringency of resolution = 3
#  due to negative effect of tumor cells on scoring
  resolution = opt$res[opt$score == max(opt$score)])

# Save a stopping point - merged and harmony aligned
qs::qsave(comb, "comb.qs")
```

The above saved qs object serves as the input for the "Primary-Cluster.Rmd"
and other downstream scripts.