---
title: "Coder Upgrade 2024: Multiomics analysis"
author: "Matt Cannon and Matt Gust"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output:
    html_document:
        toc: true
        toc_float: true
        toc_depth: 5
        number_sections: false
        code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    echo = TRUE,
    cache = TRUE,
    cache.lazy = FALSE
)
```

# Load libraries
```{r libraries, cache=FALSE, warning=FALSE, error=FALSE, message=FALSE}
library(rrrSingleCellUtils)
library(tidyverse)
library(Seurat)
library(Signac)
library(patchwork)
library(ensembldb)
```

--------------------------------------------------------------------------------

# Section 1: Overview

## What is multiomics single-cell data?

Any technology that produces multiple type of data from single cells

Each cell is profiled for multiple types of data

- RNA / ATAC
- RNA / ChIP
- RNA / Protein
- RNA / DNA
- RNA / DNA methylation / Chrom. accessibility

## Benefits of multiomics

- Direct association within individual cells between different types of data
- Avoids shakey correlations made using bulk methods on heterogeneous populations
- Can generate insights not possible with other methods

## Challenges of multiomics

- Generally, the data quality will be much worse for each individual cell compared to bulk methods
- Sample/library preparation is more complex
    - Especially for homebrew methods
- $$$
- Analysis is going to be much more complex
    - Need to analyze multiple single-cell data types
    - Need proper statistical methods to integrate the data
    - Methods are generally less mature than more commonly used methods

## 10x multiomics kit

- Generates both RNA and ATAC data from single nuclei
- Relies on cleanly isolated and fixed nuclei
    - Get only RNAs from the nucleus
    - Nuclei prep can be quite challenging
    - Likely more prone to cell-cell cross contamination

## 10x multiomics data requirements

RNA and ATAC have different sequencing requirements

- RNA
    - R1: 28bp <- has UMI and nucleus barcode
    - R2: 90bp <- has RNA sequence
    - I1/I2: 10bp <- have sample barcodes
- ATAC
    - R1: 50bp <- has genomic sequence
    - R2: 24bp <- has nucleus barcode
    - R3: 49bp <- has genomic sequence
    - I1: 8bp <- has sample barcode
- Recommended 20,000 read pairs / nucleus for GEX
    - In 10x demo dataset, 122,335 reads per nucleus
- Recommended 25,000 read pairs / nucleus for ATAC
    - In 10x demo datasets, 95,897 reads per nucleus
- Aim higher if possible
    - For 2000 cells, 100k + 100k reads per sample
        - 400M reads per sample
        - ~4 samples on a NovaSeq S1

## Data

Depending on what data type you receive you may need to do different things

- BCLs
    - Run cellranger-arc mkfastq
    - If both GEX and ATAC run on single lane, need to specify read lengths for each
    - If run seperately on different lanes, can just run mkfastq on each lane
- Fastqs
    - Run cellranger-arc count
    - Simple way is to do one run per sample
        - Input is csv with sample info
    - Output is a folder with a bunch of files/folders
        - https://support.10xgenomics.com/single-cell-multiome-atac-gex/software/pipelines/latest/using/count
        - web_summary.html              - html report
        - filtered_feature_bc_matrix/   - contains the GEX data
        - filtered_feature_bc_matrix.h5 - h5 file with both GEX and ATAC data
        - atac_peaks.bed                - ATAC peaks
        - atac_fragments.tsv.gz         - ATAC fragments
        - cloupe.cloupe                 - Loupe browser file
        - *.bam                         - BAM files

--------------------------------------------------------------------------------

## Using Seurat and Signac for 10x multiomics data individually

You essentially get two independent datasets back (GEX and ATAC) that can be analyzed independently.

The GEX data can be analyzed like normal

- One exception - if you read in your data using `Read10X_h5()`, it returns a list of two objects:
    - Gene Expression
    - Peaks
- Need to pull the GEX data out of the list and use that for analysis

I'm not going to walk through the GEX analysis as that's been covered in previous sessions

```{r}
data_path <- "~/data/input/multiomics/lesson_data/"
session_path <- "Sessions/adv_thursday/"

h5_data <-
    Read10X_h5(paste0(
        data_path,
        "human_brain_3k_filtered_feature_bc_matrix.h5"
    ))

# mt_pattern <- "^MT-" because this is human data
seurat_obj <-
    CreateSeuratObject(
        counts = h5_data$`Gene Expression`,
        assay = "RNA",
        project = "10x_multiomics"
    ) %>%
    PercentageFeatureSet(
        pattern = "^MT-",
        col.name = "percent_mt_rna",
        assay = "RNA"
    )
```

## Add the ATAC data to the Seurat object
Need to combine the datasets prior to filtering

The "Peaks" slot of our `h5_data` list is a sparse matrix. We'll add an ATAC assay to our seurat object. In order to do this, we need to provide the Peaks matrix as well as the ATAC fragment file.

```{r check-sep}
head(rownames(h5_data$Peaks))
```

The format is `chr:start-end`, so we'll pass `c(":", "-")` to the sep argument.

```{r}
frag_file <-
    paste0(
        data_path,
        "human_brain_3k_atac_fragments.tsv.gz"
    )

seurat_obj[["ATAC"]] <-
    CreateChromatinAssay(
        counts = h5_data$Peaks,
        sep = c(":", "-"),
        fragments = frag_file,
        min.cells = 5
    )
```

### Add in Metadata Output by cellranger

Cellranger is nice enough to output a csv with cell-level metadata. We'll read this in and add relevant columns to our Seurat object.

There's 30 columns in the metadata file, so we're going to only add the columns that are of interest to our ATAC QC metrics.

```{r}
metadata <-
    read.table(
        file = paste0(
            data_path,
            "human_brain_3k_per_barcode_metrics.csv"
        ),
        sep = ",",
        header = TRUE,
        row.names = NULL
    ) %>%
    subset(barcode != "NO_BARCODE" &
        barcode %in% colnames(seurat_obj)) %>%
    remove_rownames() %>%
    column_to_rownames("barcode") %>%
    dplyr::select(
        atac_mitochondrial_reads,
        atac_peak_region_fragments,
        atac_fragments
    )

seurat_obj <- AddMetaData(
    object = seurat_obj,
    metadata = metadata
)
```

### Add Gene Annotation Information

This step adds information about where genes are located in the genome, which is necessary to connect our ATAC peaks to genes. You'll need to use an annotation version appropriate for the species and genome version you used to align the data to. We're going to be using EnsDb.Hsapiens.v86 since our data was aligned to hg38. You can also use a gtf from the reference your data was aligned to, but it is only really necessary to do this in the case that your reference genome is not available from EnsDb.

We also change the seqlevels to be in the format "chr1", "chr2", etc. and set the genome to "hg38" to match the rest of the data such as the rownames.


```{r}
DefaultAssay(seurat_obj) <- "ATAC"

annotations <-
    GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86::EnsDb.Hsapiens.v86)
seqlevels(annotations) <- paste0("chr", seqlevels(annotations))
genome(annotations) <- "hg38"

# This takes ~45 minutes, so I'll just read in a pre-processed object
annotations <- qs::qread("~/data/input/multiomics/lesson_data/annotations.qs")

Annotation(seurat_obj) <- annotations
```

## Activity 1

We have another dataset for cultured neurons treated with KCl, measured 1 hour post stimulation. They were aligned to the same genome as the lesson dataset.

The dataset is in ~/data/input/multiomics/downsampled_data

### 1.1

Read in the gene expression and chromatin accessibility matrices using qs::qread()*.

*instead of the matrices being given as an h5 file, they are given as the qs object named `gene_exp_peaks_list.qs`

```{r activity-1-1}

```

### 1.2

Create a Seurat object with an RNA and ATAC assay

```{r activity-1-2}

```

### 1.3

Add gene annotations to your Seurat object
You can read in pre-made annotations from ~/data/input/multiomics/lesson_data/annotations.qs
```{r activity-1-3}

```

# Filter out likely junk cells
Not going to walk through this for RNA, as it should all be review.

## QC for ATAC

ATAC seq does have a few similar QC metrics to RNA seq, as well as some additional metrics that we don't see in RNA. Therefore we're going to introduce these new metrics one by one before we filter out any cells.

### Number of ATAC Reads

This is analagous to nCount_RNA, and follows a similar-shaped distribution to nCount_RNA albeit with larger numbers.

```{r}
feature_hist(seurat_obj, features = c("nCount_RNA", "nCount_ATAC"))
```

### Nucleosome Signal

This calculates the strength of the nucleosome signal per cell. It is a ratio of mononucleosomal fragments (between 147 and 294 bp) to nucleosome-free fragments (< 147 bp). A high nucleosome signal indicates that the Tn5 likely did a poor job of transposing the open chromatin.

The function `NucleosomeSignal` calculates this for us, and stores the values in a metadata column called `nucleosome_signal`.

We generally want nucleosome signal to be less than 2, but if it's too low your nuclei prep may have been bad and your nucleosomes all fell off.

```{r add-nuc-signal}
seurat_obj <- NucleosomeSignal(seurat_obj, assay = "ATAC")
feature_hist(seurat_obj, features = "nucleosome_signal")
```

### Transcription Start Site Enrichment

Calculates the enrichment of fragments around transcription start sites. This acts as a measure of the signal-to-noise ratio in our data. In theory, the ATAC-seq signal should be enriched around transcription start sites.

From https://www.encodeproject.org/data-standards/terms/:
> "The reads around a reference set of TSSs are collected to form an aggregate distribution of reads centered on the TSSs and extending to 2000 bp in either direction (for a total of 4000bp). This distribution is then normalized by taking the average read depth in the 100 bps at each of the end flanks of the distribution (for a total of 200bp of averaged data) and calculating a fold change at each position over that average read depth. This means that the flanks should start at 1, and if there is high read signal at transcription start sites (highly open regions of the genome) there should be an increase in signal up to a peak in the middle. We take the signal value at the center of the distribution after this normalization as our TSS enrichment metric."

*Note this won't work if you haven't added gene annotations to your data; the function will have no idea where the TSS's are!*

```{r calc-tss-enrichment}
seurat_obj <- TSSEnrichment(seurat_obj, assay = "ATAC", fast = FALSE)
feature_hist(seurat_obj, features = "TSS.enrichment")
```

### Fraction of Fragments in Peaks

This is just what it sounds like: a ratio of fragments found in accessible regions of chromatin - peaks - to total fragments. This can be considered another measure of signal-to-noise, as we would expect the majority of our fragments to come from accessible chromatin.

We'll calculate this using the metadata columns we added earlier from the cellranger output.

```{r calc-ffip}
seurat_obj$ffip <-
    seurat_obj$atac_peak_region_fragments / seurat_obj$atac_fragments
feature_hist(seurat_obj, "ffip")
```

This is also achievable with the `FRiP` function, which doesn't need the cellranger metadata file (this will be useful for the activity 😉).

First we'll run `CountFragments`. This takes a path to a fragment file, and will calculate the following metrics for each droplet:

    - CB: the cell barcode
    - frequency_count: total number of fragments sequenced for the cell
    - mononucleosome: total number of fragments with length between 147 bp and 294 bp
    - nucleosome_free: total number of fragments with length <147 bp
    - reads_count: total number of reads sequenced for the cell

After running CountFragments, we can add these metrics as metadata columns. The only metric we *need* to add in order to calculate FRiP is the total number of fragments.

HOWEVER, if we're using a counts matrix from 10x then we need to multiply the reads_count column by 2:
    they count insertion sites rather than fragments, and the CountFragment function counts fragments. If you multiply the total counts by 2, that will give you the number of insertion sites per cell, which you can use in the FRIP function to compute the accurate FRIP

```{r new-frip}
total_frags <-
    CountFragments(frag_file) %>%
    dplyr::filter(CB %in% colnames(seurat_obj)) %>% # filter out barcodes not in our object #nolint
    arrange(match(CB, colnames(seurat_obj))) %>% # arrange in same order as barcodes in object #nolint
    mutate(num_insertions = frequency_count * 2) %>% # make column for total fragments
    column_to_rownames("CB")

# add this metadata to our object
seurat_obj <- AddMetaData(seurat_obj, total_frags)

# Now we'll call FRiP
seurat_obj <-
    FRiP(
        seurat_obj,
        assay = "ATAC",
        total.fragments = "num_insertions",
        col.name = "FRiP"
    )

feature_hist(seurat_obj, features = c("FRiP", "ffip"))
```

These numbers are slightly different. This is because FRiP uses the sum of counts for each cell as the total number of fragments per cell, and we filtered out some of the really sparse peaks. Therefore out estimates are lower for FRiP than ffip. The method based on the cellranger metadata is likely better.

### Plot ATAC QC Metrics

Now let's set our cutoffs for the ATAC data.

Generally we want:

-   Trim off nCount_ATAC outliers
-   Nucleosome signal < 4-6
-   TSS enrichment > 1
-   FRiP > 0.2

```{r plot-atac-qc}
cutoffs_atac <-
    tribble(
        ~feature,             ~min_val, ~max_val,
        "nCount_ATAC",        2000,     100000,
        "nucleosome_signal",  0,        4,
        "TSS.enrichment",     1,        20, 
        "ffip",               0.2,      1,
        "nFeature_ATAC",      1000,     40000
    )
feature_hist(
    seurat_obj,
    features = c(
        "nCount_ATAC",
        "nucleosome_signal",
        "TSS.enrichment",
        "ffip",
        "nFeature_ATAC"
    ),
    cutoff_table = cutoffs_atac
)
```

## QC and Process RNA

I won't go into too much detail for this since we've seen it already.

```{r}
DefaultAssay(seurat_obj) <- "RNA"
cutoffs_rna <-
    tribble(
        ~feature,         ~min_val,  ~max_val,
        "nCount_RNA",     2000,      50000,
        "nFeature_RNA",   1000,      10000,
        "percent_mt_rna", 0,         20
    )

feature_hist(
    seurat_obj,
    features = c(
        "nCount_RNA",
        "nFeature_RNA",
        "percent_mt_rna"
    ),
    cutoff_table = cutoffs_rna
)

length(Cells(seurat_obj))

seurat_obj <-
    subset(
        seurat_obj,
        subset =
            nCount_RNA     >= cutoffs_rna$min_val[1] &
            nCount_RNA     <= cutoffs_rna$max_val[1] &
            nFeature_RNA   >= cutoffs_rna$min_val[2] &
            nFeature_RNA   <= cutoffs_rna$max_val[2] &
            percent_mt_rna <= cutoffs_rna$max_val[3] &
            nCount_ATAC    >= cutoffs_atac$min_val[1] &
            nCount_ATAC    <= cutoffs_atac$max_val[1] &
            nucleosome_signal <= cutoffs_atac$max_val[2] &
            TSS.enrichment    <= cutoffs_atac$max_val[3] &
            ffip              >= cutoffs_atac$min_val[4] &
            ffip              <= cutoffs_atac$max_val[4]
    ) %>%
    NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(verbose = FALSE) %>%
    ScaleData(verbose = FALSE) %>%
    RunPCA(verbose = FALSE) %>%
    RunUMAP(
        reduction.name = "umap_rna",
        dims = 1:10,
        verbose = FALSE
    ) %>%
    FindNeighbors(dims = 1:10, verbose = FALSE) %>%
    FindClusters(resolution = 0.4, verbose = FALSE)

seurat_obj$RNA_cluster <- Idents(seurat_obj)

length(Cells(seurat_obj))
```

# Process the ATAC data
For ATAC, we do a couple of things differently

- We don't use the standard normalization, we use term frequency-inverse document frequency (TF-IDF) instead
    - "This is a two-step normalization procedure, that both normalizes across cells to correct for differences in cellular sequencing depth, and across peaks to give higher values to more rare peaks."
- We don't run PCA
    - We run singular value decomposition (SVD) on the TD-IDF matrix generated by RunTDIDF()
    - "In linear algebra, the singular value decomposition (SVD) is a factorization of a real or complex matrix. It generalizes the eigendecomposition of a square normal matrix with an orthonormal eigenbasis to any m x n matrix." - Wikipedia
    - "The combined steps of TF-IDF followed by SVD are known as latent semantic indexing (LSI), and were first introduced for the analysis of scATAC-seq data by Cusanovich et al. 2015."
    - The reduction is named "lsi"

```{r}
DefaultAssay(seurat_obj) <- "ATAC"

seurat_obj <-
    seurat_obj %>%
    RunTFIDF() %>%
    FindTopFeatures(min.cutoff = "q0") %>%
    RunSVD() %>%
    FindNeighbors(reduction = "lsi") %>%
    FindClusters(algorithm = 3) %>%
    RunUMAP(
        reduction = "lsi",
        dims = 2:30,
        reduction.name = "umap_atac"
    )

seurat_obj$ATAC_cluster <- Idents(seurat_obj)

DimPlot(seurat_obj,
    group.by = "ATAC_cluster",
    reduction = "umap_atac",
    label = TRUE,
    repel = TRUE
) |
    DimPlot(seurat_obj,
        group.by = "RNA_cluster",
        reduction = "umap_rna",
        label = TRUE,
        repel = TRUE
    )

qs::qsave(
    seurat_obj,
    paste0(
        session_path,
        "human_brain_processed.qs"
    )
)
```

# Activity 2

Now take object from the first activity and run QC steps and processing steps.

## 2.1

Calculate the nucleosome signal, transcription start site enrichment score, and the fraction of reads in peaks.

```{r activity-2}

```

## 2.2

Plot the different QC metrics, and set cutoffs for the ATAC + RNA data.

```{r activity-2-2}

```

## 2.3

Subset the seurat object by the above QC cutoffs and process the object for both the RNA and ATAC assays.

```{r activity-2-3}

```

# Further Analysis

Now that our object is QC'ed and processed, we can proceed with more in-depth analysis of our multiomic data. This will involve identifying differentially accessible peaks between computationally defined clusters, and then connecting these peaks to our transcriptional data to uncover biological meaning.

## Subset down to just RNA_cluster 0 and 1 to look at this subset

For this analysis, we will just be comparing RNA clusters 0 and 1.

After subsetting down to these clusters, we need to re-process our data, because this subset of cells will have a different set of computationally-defined clusters, for both the ATAC and RNA assays.

```{r}
DefaultAssay(seurat_obj) <- "ATAC"

# we have to reprocess because our clusters will likely change with a different set of cells
small_obj <-
    subset(seurat_obj,
        subset = RNA_cluster %in% c(0, 1)
    ) %>%
    RunTFIDF() %>%
    FindTopFeatures(min.cutoff = "q0") %>%
    RunSVD() %>%
    FindNeighbors(reduction = "lsi") %>%
    FindClusters(
        algorithm = 3,
        resolution = 0.1
    ) %>%
    RunUMAP(
        reduction = "lsi",
        dims = 2:30,
        reduction.name = "umap_atac"
    )

small_obj$ATAC_cluster <- Idents(small_obj)

atac_dimplot <-
    DimPlot(small_obj,
        group.by = "ATAC_cluster",
        reduction = "umap_atac",
        label = TRUE
    ) +
    NoLegend()

DefaultAssay(small_obj) <- "RNA"

small_obj <-
    small_obj %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA() %>%
    RunUMAP(
        reduction.name = "umap_rna",
        dims = 1:10
    ) %>%
    FindNeighbors(dims = 1:10) %>%
    FindClusters(resolution = 0.3)

small_obj$RNA_cluster <- Idents(small_obj)

rna_dimplot <-
    DimPlot(small_obj,
        group.by = "RNA_cluster",
        reduction = "umap_rna",
        label = TRUE
    ) +
    NoLegend()

rna_dimplot + atac_dimplot
ggsave("Sessions/adv_thursday/neuron_clusters_both.png",
    width = 10,
    height = 6
)

qs::qsave(small_obj, paste0(
    session_path,
    "small_obj.qs"
))
```

## Find differential accessibility between cell types

Now we'll find differentially accessible peaks. We will look at ATAC clusters 0 and 2 in our newly-processed dataset.

```{r da_peaks}
DefaultAssay(small_obj) <- "ATAC"
Idents(small_obj) <- small_obj$ATAC_cluster
da_peaks <-
    FindMarkers(small_obj,
        ident.1 = "0",
        ident.2 = "2"
    ) %>%
    rownames_to_column("region") %>%
    left_join(ClosestFeature(small_obj,
        regions = .$region
    ) %>%
        dplyr::rename(region = query_region))

head(da_peaks)
```

# Plot ATAC peaks
Let's pick out three up and three down regulated regions
```{r}
fav_peaks <-
    da_peaks %>%
    mutate(abs_logfc = abs(avg_log2FC)) %>%
    dplyr::filter(p_val_adj < 1e-20 &
        abs_logfc > 0.5 &
        gene_biotype == "protein_coding" &
        distance == 0) %>%
    group_by(avg_log2FC > 0) %>%
    arrange(desc(abs_logfc)) %>%
    slice_head(n = 2) %>%
    pull(region)

DefaultAssay(seurat_obj) <- "ATAC"
plot_name <-
    CoveragePlot(seurat_obj,
        region = fav_peaks,
        idents = c("0", "2"),
        extend.downstream = 5000,
        extend.upstream = 5000
    )

ggsave("Sessions/adv_thursday/atac_peaks.png",
    plot_name,
    width = 15,
    height = 15
)
```

## Joint analysis of multimodal data

> Show ppt of plots

### Matching cells between datasets

Compare heterogeneity of gene expression vs. chromatin accessibility

```{r}
small_obj <-
    qs::qread(paste0(
        session_path,
        "small_obj.qs"
    ))

rna_umap <-
    Embeddings(small_obj, reduction = "umap_rna") %>%
    as.data.frame() %>%
    rownames_to_column("cell") %>%
    as_tibble() %>%
    mutate(rna_cluster = as.vector(small_obj$RNA_cluster)) %>%
    dplyr::rename(
        UMAP_1 = umaprna_1,
        UMAP_2 = umaprna_2
    )

atac_umap <-
    Embeddings(small_obj, reduction = "umap_atac") %>%
    as.data.frame() %>%
    rownames_to_column("cell") %>%
    as_tibble() %>%
    dplyr::rename(
        UMAP_1 = umapatac_1,
        UMAP_2 = umapatac_2
    ) %>%
    full_join(rna_umap %>%
        dplyr::select(cell, rna_cluster))

# Merging the UMAP data
# Shifted UMAP_1 for both by offset_num to separate the datasets

offset_num <- 15

combined_umap <-
    rna_umap %>%
    mutate(
        data_type = "RNAseq",
        UMAP_1 = UMAP_1 - offset_num
    ) %>%
    bind_rows(atac_umap %>%
        mutate(
            data_type = "ATACseq",
            UMAP_1 = UMAP_1 + offset_num
        ))

draw_cluster_1 <- 1
draw_cluster_2 <- 2

# Make plot of the combined UMAP projections
ggplot(
    combined_umap,
    aes(
        x = UMAP_1,
        y = UMAP_2,
        shape = data_type
    )
) +
    geom_point(aes(color = rna_cluster == draw_cluster_1),
        alpha = 0.5
    ) +
    scale_color_brewer(
        palette = "Paired",
        name = "Data type"
    ) +
    geom_vline(
        xintercept = 0,
        linetype = "dashed",
        color = "black"
    ) +
    geom_line(
        data = combined_umap %>%
            dplyr::filter(rna_cluster == draw_cluster_1),
        aes(
            x = UMAP_1,
            y = UMAP_2,
            group = cell
        ),
        color = "black",
        alpha = 0.05
    ) +
    geom_line(
        data = combined_umap %>%
            dplyr::filter(rna_cluster == draw_cluster_2),
        aes(
            x = UMAP_1,
            y = UMAP_2,
            group = cell
        ),
        color = "red",
        alpha = 0.05
    )

ggsave("Sessions/adv_thursday/combined_umap_lines.png",
    width = 10,
    height = 6
)

# Show this another way
# Show RNA clusters on ATAC UMAP
DimPlot(
    small_obj,
    reduction = "umap_atac",
    group.by = "RNA_cluster"
)
```

### Multimodal UMAP
Better separation of cell types due to increased dimensionality of data
```{r}
small_obj <-
    small_obj %>%
    FindMultiModalNeighbors(
        reduction.list = list("pca", "lsi"),
        dims.list = list(1:20, 2:20)
    )

# build a joint UMAP visualization
small_obj <-
    small_obj %>%
    RunUMAP(
        nn.name = "weighted.nn",
        assay = "RNA",
        reduction.name = "joint_umap"
    ) %>%
    FindClusters(
        algorithm = 3,
        graph.name = "wsnn"
    )

small_obj$joint_cluster <- Idents(small_obj)

joint_umap <-
    DimPlot(small_obj,
        reduction = "joint_umap",
        label = TRUE,
        repel = TRUE,
        pt.size = 0.5,
        group.by = "joint_cluster"
    ) +
    NoLegend()

rna_umap <-
    DimPlot(small_obj,
        reduction = "umap_rna",
        label = TRUE,
        repel = TRUE,
        pt.size = 0.5,
        group.by = "RNA_cluster"
    ) +
    NoLegend()

atac_umap <-
    DimPlot(small_obj,
        reduction = "umap_atac",
        label = TRUE,
        repel = TRUE,
        pt.size = 0.5,
        group.by = "ATAC_cluster"
    ) +
    NoLegend()

joint_umap + (rna_umap / atac_umap)
ggsave("Sessions/adv_thursday/joint_umap.png",
    width = 15,
    height = 10
)
```

### Correlation of ATAC peaks with gene expression
```{r}
DefaultAssay(small_obj) <- "ATAC"
Idents(small_obj) <- "ATAC_cluster"

small_obj <-
    RegionStats(
        small_obj,
        genome = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38
    )
# ignore the warning - our object just doesn't have some of the junk contigs

# finding differentially accessible peaks between atac cluster 0 and 2
diff_peaks <-
    FindMarkers(
        small_obj,
        assay = "ATAC",
        ident.1 = "0",
        ident.2 = "2"
    ) %>%
    rownames_to_column("region") %>%
    as_tibble() %>%
    full_join(tibble(region = rownames(small_obj))) %>%
    arrange(match(region, rownames(small_obj))) %>%
    mutate(p_val_adj = ifelse(is.na(p_val_adj), 1, p_val_adj)) %>%
    dplyr::filter(p_val_adj <= 0.05)

qs::qsave(
    diff_peaks,
    "Sessions/adv_thursday/diff_peaks.qs"
)
if (all(rownames(small_obj) == diff_peaks$region)) {
    small_obj[["ATAC"]][["neuron_signif"]] <- diff_peaks$p_val_adj < 0.05
} else {
    stop("Regions don't match")
}

# Here, RNA clusters 0 and 2 are roughly equal to ATAC clusters 0 and 2
de_genes <-
    FindMarkers(
        small_obj,
        assay = "RNA",
        ident.1 = "0",
        ident.2 = "2",
        logfc.threshold = 2
    ) %>%
    dplyr::filter(p_val_adj < 0.000001 & # Using a really low number since the p-values are so skewed
        abs(pct.1 - pct.2) > 0.6) %>%
    arrange(desc(abs(pct.1 - pct.2)))

# Getting this down to a smallish number so it doesn't take forever to run
nrow(de_genes)

# link peaks to top 100 genes as determined by the difference in pct expression
small_obj <-
    LinkPeaks(
        object = small_obj,
        peak.assay = "ATAC",
        expression.assay = "RNA",
        min.distance = 50000,
        genes.use = rownames(de_genes)[1:10]
    )
# Ignore the warnings about the junk contigs

qs::qsave(
    small_obj,
    "Sessions/adv_thursday/small_obj_linked.qs"
)

links <-
    Links(small_obj) %>%
    as.data.frame() %>%
    as_tibble() %>%
    arrange(pvalue)

links

links$peak %in% diff_peaks$region %>% summary()
# Many of these peaks associated with genes are also differentially accessible between clusters 0 and 2

CoveragePlot(
    object = small_obj,
    region = links$gene[1],
    features = links$gene[1],
    idents = c("0", "2"),
    peaks.group.by = "neuron_signif",
    expression.assay = "RNA",
    extend.upstream = 100000,
    extend.downstream = 100000
)
```

Now we're going to find enriched motifs in our differentially accessible peaks in the following steps:
    1. getMatrixSet gathers motif information matching the specified criteria
    2. AddMotifs creates an object containing DNA sequence motif information and adds it to the seurat object
    3. FindMotifs finds over-represented motifs by computing the number of differentially accessible peaks containing a motif, and compares it to all features containing that motif

Recall that the features we are calculating motif enrichment within are differentially accessible between ATAC cluster 0 and 2.

```{r}
freq_matrix <-
    TFBSTools::getMatrixSet(
        x = JASPAR2020::JASPAR2020,
        opts = list(
            collection = "CORE",
            tax_group = "vertebrates",
            all_versions = FALSE
        )
    )
# This takes a while, so lets read in the pre-made object

freq_matrix <- qs::qread("~/data/input/multiomics/lesson_data/freq_matrix.qs")

small_obj <-
    AddMotifs(
        object = small_obj,
        genome = BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38,
        pfm = freq_matrix
    )
# This takes a while, so lets read in the pre-made object

small_obj <-
    qs::qread("~/data/output/rdata_premade/multiomics_small_obj_motifs.qs")

matching_motifs <-
    FindMotifs(
        object = small_obj,
        features = links$peak %>% unique()
    )

dplyr::select(
    matching_motifs,
    percent.observed,
    percent.background,
    motif.name,
    p.adjust
) %>%
    arrange(p.adjust) %>%
    head(n = 15)
```

# Activity 3

I'm thinking for this we'll do...
    - differential accessibility between two groups in our dataset
    - get top 5 upregulated regions for each group
    - create ATAC feature plot of these regions
    - create joint umap
    - link peaks to genes
    - plot expression of peak-linked genes

## 3.1

Read in the object `input/multiomics/activity_data/activity_small.qs`. This is the object from before, downsampled to only contain cells in ATAC clusters 0 and 1. Find the top 5 differentially accessibile peaks between the two groups as determined by log-fold change

```{r activity-3-1}

```

## 3.2

Create a coverage plot of these regions

```{r activity-3-2}

```

## 3.3

Identify what genes are associated with these peaks

```{r activity-3-3}

```

## 3.4

Find what motifs are enriched in our differentially accessible peaks.

```{r activity-3-4}

```

# Resources
Methods and applications for single-cell and spatial multi-omics: https://www.nature.com/articles/s41576-023-00580-2

Into the multiverse: advances in single-cell multiomic profiling: https://www.cell.com/trends/genetics/fulltext/S0168-9525(22)00077-4

https://www.10xgenomics.com/products/single-cell-multiome-atac-plus-gene-expression

https://stuartlab.org/signac/

Merging ATAC datasets: https://stuartlab.org/signac/articles/merging.html
